### PR TITLE
Support DOTNET_ prefix

### LINKED ISSUE #47283: Can we rename COMPlus_ to DOTNET_ for environment variables?
We have a lot of configuration possible via environment variables that begin with the prefix "COMPlus", e.g. all of the values in https://github.com/dotnet/runtime/blob/16c48d6a45d4d706c327fc840e1a5d450fb189ae/src/coreclr/inc/clrconfigvalues.h.

This is 20-year old legacy finding its way into user-visible configuration for the platform.

Can we rename this to "DOTNET"?

If there are concerns about tools that have hardcoded "COMPlus", we could support both for a few releases and then delete the "COMPlus" support when the majority of hardcodings have switched over.

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Fixes https://github.com/dotnet/runtime/issues/47283

This enables support for using the `DOTNET_` prefix and is now the preferred first choice. The `COMPlus_` is still supported but is the fallback.

~~A new flag was created to ignore the `DOTNET_` prefix in case external users have created a conflicting variable name.~~
 - ~~`DOTNET_EnableDOTNETPrefix=0`~~
 - ~~`COMPlus_EnableDOTNETPrefix=0`~~

### CODE REVIEW

**Comment by jkotas** [comment]
> A new flag was created to ignore the DOTNET_ prefix in case external users have created a conflicting variable name.

I would not bother with this. If we find that there is a conflicting name, we should just fix it.

File: src/coreclr/inc/clrconfigvalues.h [hunk 1]
```diff
@@ -442,7 +438,7 @@ CONFIG_DWORD_INFO(INTERNAL_MD_MiniMDBreak, W("MD_MiniMDBreak"), 0, "ASSERT when
 CONFIG_DWORD_INFO(INTERNAL_MD_PreSaveBreak, W("MD_PreSaveBreak"), 0, "ASSERT when calling CMiniMdRw::PreSave")
 CONFIG_DWORD_INFO(INTERNAL_MD_RegMetaBreak, W("MD_RegMetaBreak"), 0, "ASSERT when creating RegMeta class")
 CONFIG_DWORD_INFO(INTERNAL_MD_RegMetaDump, W("MD_RegMetaDump"), 0, "Dump MD in 4 functions (?)")
-RETAIL_CONFIG_STRING_INFO_EX(EXTERNAL_DOTNET_MODIFIABLE_ASSEMBLIES, W("DOTNET_MODIFIABLE_ASSEMBLIES"), "Enables hot reload on debug built assemblies with the 'debug' keyword", CLRConfig::LookupOptions::DontPrependCOMPlus_ | CLRConfig::LookupOptions::TrimWhiteSpaceFromStringValue);
+RETAIL_CONFIG_STRING_INFO_EX(EXTERNAL_DOTNET_MODIFIABLE_ASSEMBLIES, W("DOTNET_MODIFIABLE_ASSEMBLIES"), "Enables hot reload on debug built assemblies with the 'debug' keyword", CLRConfig::LookupOptions::DontPrependPrefix | CLRConfig::LookupOptions::TrimWhiteSpaceFromStringValue);
```
Comment (jkotas) [inline src/coreclr/inc/clrconfigvalues.h:441]:
```suggestion
RETAIL_CONFIG_STRING_INFO_EX(EXTERNAL_DOTNET_MODIFIABLE_ASSEMBLIES, W("MODIFIABLE_ASSEMBLIES"), "Enables hot reload on debug built assemblies with the 'debug' keyword",  CLRConfig::LookupOptions::TrimWhiteSpaceFromStringValue);
```
This will allow the fast lookup path to kick in.

Comment (jkotas) [reply src/coreclr/inc/clrconfigvalues.h:441]: (not commenting on all places that can use this)

File: src/coreclr/dlls/mscordbi/CMakeLists.txt [hunk 2]
```diff
@@ -102,7 +102,17 @@ elseif(CLR_CMAKE_HOST_UNIX)
     list(APPEND COREDBI_LIBRARIES
         mdhotdata_full
         palrt
-        # share the PAL in the dac module
+    )
+
+    # Due to the pal_redefines_file adding redirects
```
Comment (jkotas) [inline src/coreclr/dlls/mscordbi/CMakeLists.txt:107]: Why is this change needed?

**Review by jkotas — APPROVED** [review]
Thank you!

File: src/coreclr/inc/clrconfigvalues.h (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/inc/clrconfigvalues.h:441]: I was going to do that, there are two others in there that could also be update, but it does change the semantics slightly because now `COMPlus_MODIFIABLE_ASSEMBLIES` is valid. Any concerns? If not I will do all of them.

File: src/coreclr/dlls/mscordbi/CMakeLists.txt (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/dlls/mscordbi/CMakeLists.txt:107]: Because for some inscrutable reason the exports in question don't exist on Linux... Oh gosh. I just realized why... I might be able to fix this. Thank you for asking!

File: src/coreclr/inc/clrconfigvalues.h (hunk 1 above)
Comment (jkotas) [reply src/coreclr/inc/clrconfigvalues.h:441]: I do not think it is a problem.

**Comment by lambdageek** [comment]
Should we reserve the `DOTNET_` prefix for variables that are accepted by both runtimes, and continue using `COMPlus_` for CoreCLR (and the BCL) and `MONO_` for MonoVM?

**Comment by AaronRobinsonMSFT** [comment]
@lambdageek That is an interesting split. @stephentoub, who filed the issue, thoughts?

**Comment by stephentoub** [comment]
> Should we reserve the DOTNET_ prefix for variables that are accepted by both runtimes, and continue using COMPlus_ for CoreCLR (and the BCL) and MONO_ for MonoVM?

I would love for us to never utter "COMPlus" again :)

If we feel we need to differentiate some values based on runtime, I could swallow having some be prefixed with CORECLR_ and some with MONO_, and then DOTNET_ for the general "applies everywhere" case.  However, I worry such a split might lead to headaches over time, as settings initially only supported by one runtime shift to being supported everywhere (which we would generally hope to be the case as we further unify and smooth over differences).  I also think it'd be potentially confusing for folks to learn about multiple prefixes and remember when to use which.  And, there are other splits I don't necessarily want to change the prefix for, e.g. settings that only make sense on one OS and that are simply ignored on others.

**Comment by lambdageek** [comment]
Yea I think in the long term it will be bad to have multiple flag prefixes as we continue to converge.  I don't, for the record, feel that it's a particularly bad idea to shift CoreCLR over to a `DOTNET_` prefix.

I'm just apprehensive bug reports like "Setting `DOTNET_GC_XYZ` doesn't do anything on Android".  But on the other hand that's really no different from "`COMPlus_GC_XYZ` doesn't do anything on Android" - either way we will need to do some work in Mono, or educate the customer, or both.

**Comment by AaronRobinsonMSFT** [comment]
> either way we will need to do some work in Mono, or educate the customer, or both.

@lambdageek This is spot on. I am going to be writing up some official doc on these configs. The past week has been an amazing journey and this system is interesting to say the least. I am hoping the education can be as simple as "go to &lt;URL&gt; and read".

**Comment by AaronRobinsonMSFT** [comment]
crossgen2smoke failures are #50539
